### ISSUE TITLE
Problem using reprosum variables

### ISSUE THREAD

**Issue by cacraigucar** [ESCOMP/CAM#335]
When creating cam6_3_009, there were some abnormalities seen when using:
reprosum_diffmax=1.0e-14
reprosum_recompute=.true.

The test was the: 
ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd

The version of CTSM being used gave different behavior for this test when using the two reprosum namelist variables

**Comment by cacraigucar** [comment]
Email from Bill Sacks which summarizes the problem.  It lists the results from Chris and Cheryl as well as the details from his CTSM testing.

First, the TL;DR:

The baseline differences can be attributed to the user_nl_cpl changes introduced in a recent CAM version (setting reprosum_diffmax and reprosum_recompute): when I remove those changes, the test returns to being bit-for-bit between CAM6_3_009 and CAM6_3_000. Therefore, if others agree, I am comfortable setting aside these answer changes as expected.
It seems like there really may be an issue with either restarts, processor-count changes or general reproducibility with this configuration – i.e., a failure in the base vs. rest comparison of the ERP test. I don't have a good feeling for the expectations, but I personally feel this warrants more investigation, if the CAM team feels it's important that this configuration be reproducible. It might be helpful to run a longer test (say, 10 days rather than 9 time steps): the reproducibility issue might show its face more reliably in a longer test like that.
Recap of what we already knew

Based on Cheryl's and Chris's testing, we knew that:

CAM6_3_009 vs. CAM6_3_000 was bit-for-bit when using ctsm1.0.dev107
CAM6_3_009 vs. CAM6_3_000 had answer changes using ctsm5.1.dev019
The rest vs. base comparison of the ERP test failed when using CAM6_3_009 with ctsm1.0.dev107, but (I think) not when using CAM6_3_000 (with either version of CTSM) or when using ctsm5.1.dev019
More details of what I tried and what I found

Finding the CTSM tag where the behavior changes

I started by doing the CAM6_3_009 vs. CAM6_3_000 comparison with various versions of CTSM. This was all in the context of the current CESM master – i.e., I checked out CESM master then just changed the cam and clm externals.

I checked a number of CTSM versions up to and including ctsm1.0.dev013; these all had:

FAIL ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd COMPARE_base_rest
PASS ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd BASELINE cam000_ctsm0xx:

(for the appropriate baseline, generated with the same ctsm tag).

However, starting with ctsm1.0.dev014 (which was a climate-changing CTSM tag), the behavior changed to:

PASS ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd COMPARE_base_rest
FAIL ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd BASELINE cam000_ctsm014: DIFF

That behavior persisted for the following tags (I didn't check every tag, but checked ctsm1.0.dev015 and ctsm1.0.dev018, in addition to ctsm1.0.dev019).

The interesting finding for me was that the baseline differences began with the same CTSM tag where the COMPARE_base_rest began to pass. This made me start to think that the two may be related to some extent.

Reverting CAM's recent change to the outfrq9s_sd testmod

One difference I noticed between CAM6_3_000 and CAM6_3_009 was that the latter has a user_nl_cpl file in the outfrq9s_sd testmod directory, containing:

reprosum_diffmax=1.0e-14
reprosum_recompute=.true.

I don't understand exactly what those settings do, but I do see that they change the drv_in file, where the original values for those variables were:

reprosum_diffmax = -1.0e-8
reprosum_recompute = .false.

So I tried running with CAM6_3_009 and ctsm1.0.dev014 but with that user_nl_cpl change removed.

Interestingly, now this brought us back to:

FAIL ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd COMPARE_base_rest
PASS ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd BASELINE cam000_ctsm014:

However, when I ran this again, the COMPARE_base_rest passed, as did the baseline comparison. I don't think this is relevant, but I'll mention: The first time, when I got the failure, I had rerun it from an existing test directory, after commenting out the user_nl_cpl settings in that test directory; the second time, when everything passed, it was with a fresh test.

In addition, I then ran a fresh test from CAM6_3_009 and ctsm1.0.dev019, with the user_nl_cpl changes removed, and got:

PASS ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd COMPARE_base_rest
PASS ERP_D_Ln9.f19_f19_mg17.FSD.cheyenne_intel.cam-outfrq9s_sd BASELINE cam000_ctsm019:

My first conclusion from this is that the baseline differences with newer versions of CTSM are attributable to the new reprosum settings in CAM's outfrq9s_sd testmod. Though I don't completely understand the impact of these settings, it doesn't seem too surprising to me that this change would change answers, and that these answer changes might only show up in certain circumstances. I could believe that the answer changes in ctsm1.0.dev014 put CAM into a different state where these reprosum changes now got activated in this test. Therefore, if others agree, I am comfortable setting aside these answer changes as expected.

My second conclusion, though, is that there really may be an issue with exact restarts and/or changing processor counts and/or possibly reproducibility as a whole with this FSD configuration. It's hard to pin it down, because these failures were not themselves reproducible, but I saw consistent failures in COMPARE_base_rest using CAM6_3_009 with ctsm5.1.dev013 and earlier, and one instance of a failure with ctsm5.1.dev014 when I removed the user_nl_cpl changes in this test. I personally feel this warrants more investigation, if the CAM team feels it's important that this configuration be reproducible. It might be helpful to run a longer test (say, 10 days rather than 9 time steps): the reproducibility issue might show its face more reliably in a longer test like that.

Bill
